### PR TITLE
adds TLS signature algorithms list feature

### LINKED ISSUE #24535: Retrieve the names of available signature algorithms in the age of providers
As suggested in the discussion below, this is to propose a new openssl app, or extension to an existing app like `list` or an external script facilitating the retrieval of currently available signature algorithms in a form suitable for consumption by the openssl config file.

### Discussed in https://github.com/openssl/openssl/discussions/24522

<div type='discussions-op-text'>

<sup>Originally posted by **baentsch** May 29, 2024</sup>
The [provider mechanism](https://www.openssl.org/docs/man3.3/man7/provider.html) allows (potentially many) different signature algorithms to be accessible to the `openssl` TLS logic. The [openssl config](https://www.openssl.org/docs/manmaster/man3/SSL_CONF_cmd.html) mechanism allows the ones actually active to be configured/limited, e.g., to [work around server problems with too many active signature algorithms](https://github.com/open-quantum-safe/oqs-provider/issues/399).

The question now is: Is there a mechanism by which the unwary user can retrieve a list of all permitted values to the "SignatureAlgorithms" config parameter, e.g., along the lines of `openssl list -signature-algorithms` (which does take providers into account, but does not output "fully qualified" algorithm names for use by the configuration, e.g. and incl. hash algs)?</div>

### PR DESCRIPTION
**Pull request by baentsch**
Fixes #24535 

This works with providers that support the "TLS-SIGALG" [capability](https://www.openssl.org/docs/manmaster/man7/provider-base.html).

This then begs the question whether this capability should be added to the built-in providers? As part of this PR or a separate one? Or would anyone prefer extending this PR with a hack to expose the current [sigalg_lookup_tbl](https://github.com/openssl/openssl/issues/24535#issuecomment-2141426971)?

### CODE REVIEW

**Comment by t8m** [comment]
I think it should list the built-in sig algs as well (somehow separated from the provider ones). It is similar to the `
openssl list -cipher-algorithms` command.

**Comment by baentsch** [comment]
> I think it should list the built-in sig algs as well (somehow separated from the provider ones). It is similar to the ` openssl list -cipher-algorithms` command.

I'd agree. The question is how it should do that. I see two options:
1) Expose [sigalg_lookup_table](https://github.com/openssl/openssl/blob/d8def79838cd0d5e7c21d217aa26edb5229f0ab4/ssl/t1_lib.c#L1383) (as mentioned by @mattcaswell ) or
2) Expose built-in providers' sigalgs via capabilities (sounds like having a long-term benefit -- but IMO warrants a separate PR).

Any preferences or alternatives?

**Comment by mattcaswell** [comment]
This PR looks good.

I agree there are long term benefits in having the built-in sigalgs exposed via the standard providers as a capability. We perhaps should have done so from the beginning. However this does bring the disadvantage that it only works for those providers where we have made this change. The most obvious case where this would not be true is if someone is using a validated version of the fips provider from an older release (e.g. 3.0.9) with a more recent version of OpenSSL. We could perhaps cover that via documentation (e.g. basically say that it may not correctly report sigalgs for older providers)

**Comment by baentsch** [comment]
> We perhaps should have done so from the beginning. 

Mea culpa; I just didn't think about the built-in providers when adding provider-loaded TLS sigalgs (they worked :). Speaks for back-porting this as well as a "TLS-SIGALG" capabilities enhancement PR for the built-in providers, though.

> The most obvious case where this would not be true is if someone is using a validated version of the fips provider from an older release (e.g. 3.0.9) with a more recent version of OpenSSL.

Well, such FIPS provider would (could! should?) not support provider-loaded sigalgs anyway, so that'd be no problem, right? Or would the FIPS-provider-3.0.9-based TLS sigalg work even though it'd not be reported ("courtesy" `sigalg_lookup_tbl`?). If so (?) that'd speak for exposing `sigalg_lookup_tbl` to this new `list` app command.

**Comment by mattcaswell** [comment]
> Well, such FIPS provider would (could! should?) not support provider-loaded sigalgs anyway, so that'd be no problem, right? Or would the FIPS-provider-3.0.9-based TLS sigalg work even though it'd not be reported ("courtesy" sigalg_lookup_tbl?). If so (?) that'd speak for exposing sigalg_lookup_tbl to this new list app command.

Right - a 3.0.9 based FIPS provider should "just work" for TLS because the sigalg_lookup_tbl provides all the necessary codepoints - but the new list feature would not work with it because it doesn't have the "TLS-SIGALG" capability. This was my point above about possibly handling this via the list app documentation, i.e. "we could just say in the docs: don't expect this to work for old providers"

We could maybe do both (1) and (2), i.e. the motivation for (1) would be to ensure that the list command works no matter whether you are using an old provider or not. The motivation for (2) would be to get the long term benefits of having these providers declaring these things.

**Comment by t8m** [comment]
IMO we should do (1) as the first step. (2) can be done independently any time later.

**Comment by baentsch** [comment]
(1) implemented in latest commit. A bit more complicated than I thought because `sigalg_lookup_tbl` didn't contain all algorithm names; I found them in a "mirror data structure" https://github.com/openssl/openssl/blob/42230f294ae97cbd50052038499e091d0060ba8e/ssl/t1_trce.c#L570 though: Hoping this approach is OK.

**Comment by baentsch** [comment]
Hmm -- running into all kind of "style" mistakes (incl. in code I didn't change): Is there a place where I could read up on this and learn how to run & fix this (new?) test locally instead of via "CI-debugging"? Thanks in advance for a pointer!

**Comment by nhorman** [comment]
If you rebase on top of master, many should be resolved.  Additionally you may want to pull in this commit
https://github.com/openssl/openssl/pull/24861/commits/cdca0cd04b6761ece515796b7f19d536f8a8a225

As it fixes the REND problems that arise when git diff sometimes doesn't produce a proper git end line count

For now we can waive the style issues by setting the style:waived label

**Comment by baentsch** [comment]
Thanks for this quick reaction andf guidance, @nhorman !

> Please try to parse out any relevant style issues relevant to your patch set

Would do -- but how do I trigger local style testing? With the label now set I don't see the CI error messages any more...

Another question pertaining to a (presumably ssl-lib local) symbol I introduced to expose t1_trce functionality to t1_lib: This seems to work for most setups, but not for two CI jobs: Any idea what I'd need to change there? Do I need to also declare `ssl_get_sigalg_name` as a public symbol now?

**Comment by nhorman** [comment]
No worries.  If you examine the ci job, you'll see that if you click the triangle next to the commit range error report, the style errors are still actually reported, the setting of the style:waived flag just forces the ci check to return successfully.  So you can still review them at will.

As for your symbol question, I'm not sure.  I'll need to look at the PR.to tell more (in the car atm), but what you're describing sounds like a static vs dynamic issue.  If you create a symbol and don't export it, it doesn't show up in the dao file, but if you link statically, you can still resolve the symbol because you have visibility on all symbols in the libssl.a and libcrypro.a files

Does that sound like what might be going on?

**Comment by baentsch** [comment]
> So you can still review them at will.

Thanks for the hint: Indeed all code lines changed by the PR are now accepted, so I consider this dealt with.

> if you link statically, you can still resolve the symbol because you have visibility on all symbols in the libssl.a and libcrypro.a files
> Does that sound like what might be going on?

Nope, to the contrary: The problematic one is indeed a config ("minimal") that generates .a files. All .so configs pass. Together with a Windows job it's the only one complaining about a (new) symbol that is (within) "ssl_local". I'm anyway not happy about this as this is just a workaround for missing algorithm names in `sigalg_lookup_tbl`: I'd much rather change that table to contain the names instead but don't know whether this (`lu->name == NULL` for some sigalgs) is somehow intentional (?).

File: apps/list.c [hunk 1]
```diff
@@ -1576,7 +1627,7 @@ int list_main(int argc, char **argv)
         switch (o) {
         case OPT_EOF:  /* Never hit, but suppresses warning */
         case OPT_ERR:
-opthelp:
+        opthelp:
```
Comment (paulidale) [inline apps/list.c:1630]: nit: don't indent this please.

**Comment by paulidale** [comment]
OTC: **back port through to 3.2 or not?**

File: apps/list.c (hunk 1 above)
Comment (baentsch) [reply apps/list.c:1630]: That was not my choice: I did this only to "satisfy" the style-checker tool in CI. Will revert as part of PR review update now that style:waived tag is set and further feedback is received.

**Comment by t8m** [comment]
> Nope, to the contrary: The problematic one is indeed a config ("minimal") that generates .a files. All .so configs pass. Together with a Windows job it's the only one complaining about a (new) symbol that is (within) "ssl_local". I'm anyway not happy about this as this is just a workaround for missing algorithm names in `sigalg_lookup_tbl`: I'd much rather change that table to contain the names instead but don't know whether this (`lu->name == NULL` for some sigalgs) is somehow intentional (?).

You would have to factorize the function out of t1_trce.c as that is not built in builds disabling the SSL trace support. But can't SSL_group_to_name() be used to convert  a group id to name?

**Comment by baentsch** [comment]
> But can't SSL_group_to_name() be used to convert a group id to name?

Thanks for that suggestion; this call seems to require an SSL connection (object), though. Is your idea to create a "dummy" SSL object without real connection for this purpose? OK, let me try this...

**Comment by baentsch** [comment]
> You would have to factorize the function out of t1_trce.c as that is not built in builds disabling the SSL trace support.

Back to that option. OK to move this function (and the related data structs) to t1_lib then?

**Comment by baentsch** [comment]
On third thought: Instead of introducing another function wouldn't it be much cleaner to replace the few NULL `name` entries in https://github.com/openssl/openssl/blob/01753c09bbfdffcefd555b4c21e50e68af346129/ssl/t1_lib.c#L1383-L1483 with the names properly listed in https://github.com/openssl/openssl/blob/01753c09bbfdffcefd555b4c21e50e68af346129/ssl/t1_trce.c#L570-L602, e.g., by way of #defines's. Does anything speak against this?

**Comment by t8m** [comment]
> > But can't SSL_group_to_name() be used to convert a group id to name?
> 
> Oops -- group_to_name won't resolve sig alg names.

Ouch, I'm sorry, not sure how I confused these two. :grinning:

**Comment by t8m** [comment]
> On third thought: Instead of introducing another function wouldn't it be much cleaner to replace the few NULL `name` entries in
> 
> https://github.com/openssl/openssl/blob/01753c09bbfdffcefd555b4c21e50e68af346129/ssl/t1_lib.c#L1383-L1483
> with the names properly listed in
> 
> https://github.com/openssl/openssl/blob/01753c09bbfdffcefd555b4c21e50e68af346129/ssl/t1_trce.c#L570-L602
> , e.g., by way of #defines's. Does anything speak against this?

Adding the name there will make it possible to use the name in the SignatureAlgorithms config setting. However I do not see any problem with that. @mattcaswell any idea why are some of these names NULL?

File: apps/list.c (hunk 1 above)
Comment (baentsch) [reply apps/list.c]: Changed back in latest commit.

**Comment by baentsch** [comment]
> why are some of these names NULL?

Latest commit changes this and streamlines use of these names within t1_lib and t1_trce, doing away with the need for a helper function.

Note: [Here's](https://github.com/openssl/openssl/blob/a753547eefc9739f341824a0cb0642afe7a06fcc/apps/lib/s_cb.c#L753) another duplication of these names: Worthwhile streamlining, too? Same or separate PR?

**Comment by t8m** [comment]
> OTC: **back port through to 3.2 or not?**

**OTC: This is a new feature. There is no compelling reason to justify backporting it to 3.2 in exception of the stable branch policy.**

**Comment by baentsch** [comment]
> > OTC: **back port through to 3.2 or not?**
> 
> **OTC: This is a new feature. There is no compelling reason to justify backporting it to 3.2 in exception of the stable branch policy.**

ACK. Rebased & ready for review.

File: ssl/t1_lib.c [hunk 2]
```diff
@@ -1594,6 +1594,66 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
     return ret;
 }
 
+#define SIGLEN_BUF_INCREMENT 100
+
+char *SSL_get_builtin_sigalgs(OSSL_LIB_CTX *libctx)
```
Comment (t8m) [inline ssl/t1_lib.c:1599]: Given the function returns an allocated memory that needs to be freed, please rename it to SSL_get1_builtin_sigalgs().

File: ssl/t1_lib.c [hunk 3]
```diff
@@ -1594,6 +1594,66 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+            const char *sa = lu->name;
+
+            if (sa != NULL) {
+                if (strlen(sa) + strlen(retval) + 1 >= maxretlen) {
+                    maxretlen += SIGLEN_BUF_INCREMENT;
+                    retval = OPENSSL_realloc(retval, maxretlen);
+                }
+                if (retlen > 0)
+                    OPENSSL_strlcat(retval, ":", ++retlen);
+                retlen += strlen(sa)+1;
```
Comment (t8m) [inline ssl/t1_lib.c:1643]: Nit: spaces around `+`

File: include/openssl/tls1.h [hunk 4]
```diff
@@ -282,6 +282,8 @@ int SSL_get_sigalgs(SSL *s, int idx,
                     int *psign, int *phash, int *psignandhash,
                     unsigned char *rsig, unsigned char *rhash);
 
+char * SSL_get_builtin_sigalgs(OSSL_LIB_CTX *libctx);
```
Comment (t8m) [inline include/openssl/tls1.h:285]: Nit: no space after `*`

File: util/other.syms [hunk 5]
```diff
@@ -834,3 +834,4 @@ EVP_PKEY_base_id                        define
 SSL_set_retry_verify                    define
 TS_VERIFY_CTX                           define
 CMAC_CTX                                define
+SSL_get_builtin_sigalgs                 define
```
Comment (t8m) [inline util/other.syms:837]: This line should be removed as it is not a define

File: doc/man3/SSL_get_builtin_sigalgs.pod [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+=head1 NAME
+
+SSL_get_builtin_sigalgs - get list of built-in signature algorithms
+
+=head1 SYNOPSIS
+
+ #include <openssl/tls1.h>
+
+ char * SSL_get_builtin_sigalgs(OSSL_LIB_CTX *libctx);
```
Comment (t8m) [inline doc/man3/SSL_get_builtin_sigalgs.pod:11]: Nit: no space after `*`

File: doc/man3/SSL_get_builtin_sigalgs.pod [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+=head1 DESCRIPTION
+
+Return the colon-separated list of built-in and available TLS signature
+algorithms.
+The string returned must be freed by the user using L<OPENSSL_free(3)>.
+
+
```
Comment (t8m) [inline doc/man3/SSL_get_builtin_sigalgs.pod:19]: Nit: please drop this extra empty line

File: ssl/t1_lib.c [hunk 8]
```diff
@@ -1594,6 +1594,66 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
 }
 
+#define SIGLEN_BUF_INCREMENT 100
+
+char *SSL_get_builtin_sigalgs(OSSL_LIB_CTX *libctx)
+{
+    size_t i, retlen = 0, maxretlen = SIGLEN_BUF_INCREMENT;
+    const SIGALG_LOOKUP *lu;
+    EVP_PKEY *tmpkey = EVP_PKEY_new();
+    char *retval = OPENSSL_zalloc(maxretlen);
```
Comment (t8m) [inline ssl/t1_lib.c:1604]: Please check the return value for NULL.

File: ssl/t1_lib.c [hunk 9]
```diff
@@ -1594,6 +1594,66 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+            enabled = 0;
+        EVP_PKEY_CTX_free(pctx);
+
+        if (enabled) {
+            const char *sa = lu->name;
+
+            if (sa != NULL) {
+                if (strlen(sa) + strlen(retval) + 1 >= maxretlen) {
+                    maxretlen += SIGLEN_BUF_INCREMENT;
+                    retval = OPENSSL_realloc(retval, maxretlen);
```
Comment (t8m) [inline ssl/t1_lib.c:1639]: Please check the return value for NULL,  also release the original value in such case and return error (i.e. NULL)

File: ssl/t1_lib.c [hunk 10]
```diff
@@ -1594,6 +1594,66 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+                }
+                if (retlen > 0)
+                    OPENSSL_strlcat(retval, ":", ++retlen);
+                retlen += strlen(sa)+1;
+                OPENSSL_strlcat(retval, sa, retlen);
+            }
+            else {
```
Comment (t8m) [inline ssl/t1_lib.c:1646]: This should be a single line

File: ssl/t1_lib.c [hunk 11]
```diff
@@ -1594,6 +1594,66 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+        }
+
+        if (!EVP_PKEY_set_type(tmpkey, lu->sig)) {
+            enabled = 0;
+            continue;
+        }
+        pctx = EVP_PKEY_CTX_new_from_pkey(libctx, tmpkey, NULL);
+        /* If unable to create pctx we assume the sig algorithm is unavailable */
+        if (pctx == NULL)
+            enabled = 0;
```
Comment (t8m) [inline ssl/t1_lib.c:1630]: As all the errors are not fatal here, it should be inside `ERR_set_mark();` `ERR_pop_to_mark();` calls. (Also pop the mark before `continue`.)

File: apps/list.c [hunk 12]
```diff
@@ -747,6 +748,51 @@ static void list_signatures(void)
...
+    /* As built-in providers don't have this capability, never error */
+    return 1;
+}
+
+static void list_tls_signatures(void)
+{
+    int tls_sigalg_listed = 0;
+    char *builtin_sigalgs = SSL_get_builtin_sigalgs(app_get0_libctx());
+
+    if (builtin_sigalgs && strlen(builtin_sigalgs)>0) {
```
Comment (t8m) [inline apps/list.c:782]: Nits: `builtin_sigalgs != NULL && strlen(builtin_sigalgs) > 0`

File: doc/man3/SSL_get_builtin_sigalgs.pod [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+=head1 NOTES
+
+The string may be empty (strlen==0) if none of the built-in TLS signature
+algorithms can be activated, e.g., if suitable providers are missing.
+
+=head1 RETURN VALUES
+
+A newly allocated string is always returned but it may have strlen == 0.
```
Comment (t8m) [inline doc/man3/SSL_get_builtin_sigalgs.pod:27]: There might be an allocation error.

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/t1_lib.c (hunk 2 above)
Comment (baentsch) [reply ssl/t1_lib.c]: done

File: ssl/t1_lib.c (hunk 3 above)
Comment (baentsch) [reply ssl/t1_lib.c]: fixed

File: include/openssl/tls1.h (hunk 4 above)
Comment (baentsch) [reply include/openssl/tls1.h]: fixed

File: util/other.syms (hunk 5 above)
Comment (baentsch) [reply util/other.syms]: done

File: doc/man3/SSL_get_builtin_sigalgs.pod (hunk 6 above)
Comment (baentsch) [reply doc/man3/SSL_get_builtin_sigalgs.pod]: fixed

File: doc/man3/SSL_get_builtin_sigalgs.pod (hunk 7 above)
Comment (baentsch) [reply doc/man3/SSL_get_builtin_sigalgs.pod]: done

File: ssl/t1_lib.c (hunk 8 above)
Comment (baentsch) [reply ssl/t1_lib.c:1604]: blush. done.

File: ssl/t1_lib.c (hunk 9 above)
Comment (baentsch) [reply ssl/t1_lib.c:1647]: done

File: ssl/t1_lib.c (hunk 10 above)
Comment (baentsch) [reply ssl/t1_lib.c]: done

File: apps/list.c (hunk 12 above)
Comment (baentsch) [reply apps/list.c]: done

File: doc/man3/SSL_get_builtin_sigalgs.pod (hunk 13 above)
Comment (baentsch) [reply doc/man3/SSL_get_builtin_sigalgs.pod]: True. Possible return value NULL now documented.

**Review by t8m — APPROVED** [review]

File: apps/list.c [hunk 14]
```diff
@@ -747,6 +748,51 @@ static void list_signatures(void)
...
 }
 
+static int list_provider_tls_sigalgs(const OSSL_PARAM params[], void *data)
+{
+    const OSSL_PARAM *p;
+
+    /* Get registered IANA name */
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_IANA_NAME);
+    if (p != NULL && p->data_type == OSSL_PARAM_UTF8_STRING) {
+        if (*((int *)data))
```
Comment (mattcaswell) [inline apps/list.c:758]: Probably this should be:

```suggestion
        if (*((int *)data) > 0)
```

File: apps/list.c [hunk 15]
```diff
@@ -747,6 +748,51 @@ static void list_signatures(void)
...
+    /* As built-in providers don't have this capability, never error */
+    return 1;
+}
+
+static void list_tls_signatures(void)
+{
+    int tls_sigalg_listed = 0;
+    char *builtin_sigalgs = SSL_get1_builtin_sigalgs(app_get0_libctx());
+
+    if (builtin_sigalgs != NULL && strlen(builtin_sigalgs)>0) {
```
Comment (mattcaswell) [inline apps/list.c:782]:
```suggestion
    if (builtin_sigalgs != NULL && strlen(builtin_sigalgs) > 0) {
```

File: ssl/t1_lib.c [hunk 16]
```diff
@@ -1594,6 +1594,76 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+                if (strlen(sa) + strlen(retval) + 1 >= maxretlen) {
+                    maxretlen += SIGLEN_BUF_INCREMENT;
+                    retval = OPENSSL_realloc(retval, maxretlen);
+                    if (retval == NULL) {
+                        return NULL;
+                    }
+                }
+                if (retlen > 0)
+                    OPENSSL_strlcat(retval, ":", ++retlen);
+                retlen += strlen(sa) + 1;
```
Comment (mattcaswell) [inline ssl/t1_lib.c:1654]: Why +1 here?

File: ssl/t1_lib.c [hunk 17]
```diff
@@ -1594,6 +1594,76 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+                OPENSSL_strlcat(retval, sa, retlen);
```
Comment (mattcaswell) [inline ssl/t1_lib.c:1655]: Shouldn't the final arg here really by `maxretlen`, i.e. the size of the `retval` buffer. Same for the other `OPENSSL_strlcat` call above

File: ssl/t1_lib.c [hunk 18]
```diff
@@ -1594,6 +1594,76 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
 }
 
+#define SIGLEN_BUF_INCREMENT 100
+
+char *SSL_get1_builtin_sigalgs(OSSL_LIB_CTX *libctx)
+{
+    size_t i, retlen = 0, maxretlen = SIGLEN_BUF_INCREMENT;
+    const SIGALG_LOOKUP *lu;
+    EVP_PKEY *tmpkey = EVP_PKEY_new();
+    char *retval = OPENSSL_zalloc(maxretlen);
```
Comment (mattcaswell) [inline ssl/t1_lib.c:1604]: zalloc seems excessive here. It is only necessary for the first byte of the buffer to be 0.

File: ssl/t1_lib.c [hunk 19]
```diff
@@ -1594,6 +1594,76 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+        if (pctx == NULL)
+            enabled = 0;
+        ERR_pop_to_mark();
+        EVP_PKEY_CTX_free(pctx);
+
+        if (enabled) {
+            const char *sa = lu->name;
+
+            if (sa != NULL) {
+                if (strlen(sa) + strlen(retval) + 1 >= maxretlen) {
```
Comment (mattcaswell) [inline ssl/t1_lib.c:1645]: Can't we use `retlen` in place of `strlen(retval)` here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Perhaps a CHANGES.md entry is appropriate for this new feature?

File: ssl/t1_lib.c (hunk 17 above)
Comment (baentsch) [reply ssl/t1_lib.c]: Good observation. Makes the logic much simpler and removes the need for "retlen" altogether (also resolving most other comments). Addressed in latest commit.

File: apps/list.c (hunk 15 above)
Comment (baentsch) [reply apps/list.c]: fixed in latest commit

File: apps/list.c (hunk 14 above)
Comment (baentsch) [reply apps/list.c]: fixed in latest commit

**Comment by baentsch** [comment]
> Perhaps a CHANGES.md entry is appropriate for this new feature?

Not entirely sure as it's a rather small feature. Added in latest commit -- feel free to keep/amend/remove as you see fit.

File: ssl/t1_lib.c (hunk 18 above)
Comment (baentsch) [reply ssl/t1_lib.c]: changed in latest commit

File: ssl/t1_lib.c (hunk 19 above)
Comment (baentsch) [reply ssl/t1_lib.c:1648]: removed retlen

File: apps/list.c [hunk 20]
```diff
@@ -747,6 +748,51 @@ static void list_signatures(void)
...
+    /* As built-in providers don't have this capability, never error */
+    return 1;
+}
+
+static void list_tls_signatures(void)
+{
+    int tls_sigalg_listed = 0;
+    char *builtin_sigalgs = SSL_get1_builtin_sigalgs(app_get0_libctx());
+
+    if (builtin_sigalgs != NULL && strlen(builtin_sigalgs) > 0) {
```
Comment (FdaSilvaYY) [inline apps/list.c:782]: - Just testing `if (builtin_sigalgs != NULL && builtin_sigalgs[0] != 0)` is simpler IMHO.

Comment (baentsch) [reply apps/list.c]: It is. So changed. Thanks for the suggestion.

File: ssl/t1_lib.c [hunk 21]
```diff
@@ -1594,6 +1594,78 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+                    maxretlen += SIGLEN_BUF_INCREMENT;
+                    retval = OPENSSL_realloc(retval, maxretlen);
```
Comment (mattcaswell) [inline ssl/t1_lib.c:1650]: This actually leaks retval here in the case where the realloc fails and returns NULL

File: ssl/t1_lib.c [hunk 22]
```diff
@@ -1594,6 +1594,78 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
...
+                    if (retval == NULL) {
+                        return NULL;
+                    }
```
Comment (mattcaswell) [inline ssl/t1_lib.c:1653]: nit: `{}` not necessary for just a single line if

File: CHANGES.md [hunk 23]
```diff
@@ -29,6 +29,10 @@ OpenSSL 3.4
 
 ### Changes between 3.3 and 3.4 [xx XXX xxxx]
 
+ * Add feature to retrieve configured TLS signature algorithms.
```
Comment (mattcaswell) [inline CHANGES.md:32]: I'd mention something about the list app

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: CHANGES.md (hunk 23 above)
Comment (baentsch) [reply CHANGES.md]: So added in https://github.com/openssl/openssl/pull/24821/commits/bff436b74df85a6acdf34e31415a31f4fc145f21.

File: ssl/t1_lib.c (hunk 22 above)
Comment (baentsch) [reply ssl/t1_lib.c:1655]: generally changed to not leak `retval` in https://github.com/openssl/openssl/pull/24821/commits/bff436b74df85a6acdf34e31415a31f4fc145f21.

File: ssl/t1_lib.c (hunk 21 above)
Comment (baentsch) [reply ssl/t1_lib.c]: Ouch. Correct. Fixed in https://github.com/openssl/openssl/pull/24821/commits/bff436b74df85a6acdf34e31415a31f4fc145f21.

**Review by t8m — APPROVED** [review]

File: ssl/ssl_local.h [hunk 24]
```diff
@@ -2209,6 +2209,40 @@ typedef enum downgrade_en {
...
+#define TLSEXT_SIGALG_rsa_pkcs1_sha1_name                            "rsa_pkcs1_sha1"
+#define TLSEXT_SIGALG_dsa_sha256_name                                "dsa_sha256"
+#define TLSEXT_SIGALG_dsa_sha384_name                                "dsa_sha384"
+#define TLSEXT_SIGALG_dsa_sha512_name                                "dsa_sha512"
+#define TLSEXT_SIGALG_dsa_sha224_name                                "dsa_sha224"
+#define TLSEXT_SIGALG_dsa_sha1_name                                  "dsa_sha1"
+#define TLSEXT_SIGALG_gostr34102012_256_intrinsic_name               "gost2012_256"
+#define TLSEXT_SIGALG_gostr34102012_512_intrinsic_name               "gost2012_512"
+#define TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256_name       "gost2012_256"
+#define TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512_name       "gost2012_512"
```
Comment (Sashan) [inline ssl/ssl_local.h:2237]: Would it make sense to distinguish `gostr34102012_*_intrinsic` from `gostr34102012_*_gostr34112012_*`? We currently use `gost2012_*` for all of them.

I also think  `TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256_name` should be changed to `TLSEXT_SIGALG_gostr34102012_256_gostr34112012_name` (there is one more extra 256).
Same goes to `TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512_name`.

thanks.

**Review by Sashan — CHANGES_REQUESTED** [review]

**Review by mattcaswell — APPROVED** [review]
Approved, but please consider the comment above from @Sashan

File: ssl/ssl_local.h (hunk 24 above)
Comment (baentsch) [reply ssl/ssl_local.h:2237]: > Would it make sense to distinguish gostr34102012_*_intrinsic from gostr34102012_*_gostr34112012_*? We currently use gost2012_* for all of them.

Conceptually sure -- but this decision to label all "Gost's" the same [pre-dates this PR](https://github.com/openssl/openssl/blob/fd39d1c80cd5bd9cb5c64e3fc96102397e5e860f/ssl/t1_trce.c#L594-L597), so touching this seems risky as this name may be externally used/referenced. Who could comment on this mainly Gost-related question? @beldmit?

> I also think TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256_name should be changed to TLSEXT_SIGALG_gostr34102012_256_gostr34112012_name

Sure -- we can do this. But we could also use "TLSEXT_SIGALG_gostr34102012_gostr34112012_256_name" instead with the logic that the presumed strength (256) trails the presumed base algorithm name. Again, unsure what suits better the Gost spirit (sorry, couldn't resist the pun :).

Oh, and this naming again has nothing really to do with this PR but is just following what is in the current "master":

https://github.com/openssl/openssl/blob/fd39d1c80cd5bd9cb5c64e3fc96102397e5e860f/ssl/t1_trce.c#L594-L597

**Comment by baentsch** [comment]
> Approved, but please consider the comment above from @Sashan

Considered (&commented, see above) but do not think this is something this PR can/should resolve.

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the contribution.

**Pull request closed without merging**